### PR TITLE
Add support for realtime tuning kernel arguments

### PR DESCRIPTION
**Pull request by jordigilh**
This PR allows to define the amount of hugepages to be passed on to the kernel. The current implementation has the value hardcoded to 64 pages, which translate into 128Mi per node. This change is required to successfully boot up the fedora container disk used in the realtime functional tests in kubevirt, since the kernel is unable to boot with the 128Mi available in any node.

This PR adds 3 new flags to the `gocli run` cmd:
*  `hugepages-count`: Determines the number of hugepages to create. Defaults to 64. The value of this flag is captured via the environment variable `KUBEVIRT_HUGEPAGES_COUNT`
* `hugepages-2m`: When enabled, it uses `2M` page sizes. Defaults to true.
* `enable-realtime-scheduler`: When enabled, it configures each node to allow processes to be scheduled in realtime without time constrains by running the command: `sysctl kernel.sched_rt_runtime_us=-1`. Without this change, it is not possible to change the scheduling of the VM to use type FIFO and set the process priority to 1. The value of this flag is captured via the environment variable `KUBEVIRT_REALTIME_SCHEDULER`

/cc @vladikr @rmohr @davidvossel @pkliczewski

### CODE REVIEW

File: cluster-provision/centos8/scripts/kernel.args [hunk 1]
```diff
@@ -1 +1 @@
-root=/dev/vda1 ro no_timer_check console=tty0 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 elevator=noop spectre_v2=off nopti hugepagesz=2M hugepages=64 intel_iommu=on modprobe.blacklist=nouveau
\ No newline at end of file
+root=/dev/vda1 ro no_timer_check console=tty0 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 elevator=noop spectre_v2=off nopti hugepagesz=2M intel_iommu=on modprobe.blacklist=nouveau
```
Comment (rmohr) [inline cluster-provision/centos8/scripts/kernel.args:1]: Since we already take out hugepages, let's also move out the `hugepagesz=2M` setting from here and move it up to `gocli`. We have thenall parts which one would potentially want to modify removed from here.

File: cluster-provision/k8s/1.20/nodes.sh [hunk 2]
```diff
@@ -5,7 +5,10 @@ set -ex
 source /var/lib/kubevirtci/shared_vars.sh
 
 # Ensure that hugepages are there
-cat /proc/meminfo | sed -e "s/ //g" | grep "HugePages_Total:64"
+if [ -n $HUGEPAGES_COUNT ]; then
```
Comment (rmohr) [inline cluster-provision/k8s/1.20/nodes.sh:8]: Let's just remove this check. We know that we pass it to the kernel. That should be good enough. If we want to test that, I think we should do it with our conformance tests.

**Review by rmohr — CHANGES_REQUESTED** [review]
Looks very good. Two suggestions.

File: cluster-provision/centos8/scripts/kernel.args (hunk 1 above)
Comment (jordigilh) [reply cluster-provision/centos8/scripts/kernel.args:1]: Sounds good. I'll parametrize it as well. Since I don't know which kind of environments it will run (x86, arm), I will not check the value of the page size to avoid locking it to certain platforms (x86). What do you think?

Comment (rmohr) [reply cluster-provision/centos8/scripts/kernel.args:1]: > Sounds good. I'll parametrize it as well. Since I don't know which kind of environments it will run (x86, arm), I will not check the value of the page size to avoid locking it to certain platforms (x86). What do you think?

How about adding the following flag: `--hugepages-2m` instead of an extra parameter?

Comment (jordigilh) [reply cluster-provision/centos8/scripts/kernel.args:1]: That would be easier indeed. I'll just add the flag and set the default value as true for now.

This means we're skipping support for `1G` and ppc64 arch page sizes, right?  [(ppc64 uses 4k and 16M)](https://www.kernel.org/doc/html/latest/admin-guide/mm/hugetlbpage.html).

**Comment by jordigilh** [comment]
@rmohr finished implementing all changes. I've asked @vladikr for a review :smile:

**Comment by vladikr** [comment]
Thanks @jordigilh 
Roman's comments have been addressed. It looks good to me.
/approve

File: cluster-provision/gocli/cmd/run.go [hunk 3]
```diff
@@ -499,8 +519,15 @@ func run(cmd *cobra.Command, args []string) (retErr error) {
 		if len(nodeQemuArgs) > 0 {
 			additionalArgs = append(additionalArgs, "--qemu-args", shellescape.Quote(nodeQemuArgs))
 		}
-		if len(kernelArgs) > 0 {
-			additionalArgs = append(additionalArgs, "--additional-kernel-args", shellescape.Quote(kernelArgs))
+
+		var hugepagesAmount string
```
Comment (jean-edouard) [inline cluster-provision/gocli/cmd/run.go:523]: Maybe I missed something but what's the point of this variable?
Why not just use `hugepagesSize2M` directly?

File: cluster-provision/gocli/cmd/run.go [hunk 4]
```diff
@@ -706,6 +733,20 @@ func run(cmd *cobra.Command, args []string) (retErr error) {
 		}
 	}
 
+	if realtimeSchedulingEnabled {
+		nodeName := nodeNameFromIndex(1)
+		success, err := docker.Exec(cli, nodeContainer(prefix, nodeName), []string{
+			"/bin/bash",
+			"-c",
+			"ssh.sh sudo sysctl kernel.sched_rt_runtime_us=-1",
```
Comment (jean-edouard) [inline cluster-provision/gocli/cmd/run.go:741]: That won't store the value into sysctl.conf, right?
Maybe we should also do that, in case the node ever reboots?

File: cluster-provision/gocli/cmd/run.go [hunk 5]
```diff
@@ -792,7 +833,6 @@ func prepareDeviceForAssignment(cli *client.Client, nodeContainer, pciID, pciAdd
 func prepareEtcdDataMount(node string, etcdDataDir string, mountSize string) error {
 	var err error
 	var success bool
-
```
Comment (jean-edouard) [inline cluster-provision/gocli/cmd/run.go:795]: Nit: unnecessary whitespace removal

File: cluster-provision/k8s/1.19/node01.sh [hunk 6]
```diff
@@ -2,9 +2,6 @@
 
 set -ex
 
-# Ensure that hugepages are there
-cat /proc/meminfo | sed -e "s/ //g" | grep "HugePages_Total:64"
```
Comment (jean-edouard) [inline cluster-provision/k8s/1.19/node01.sh:6]: Pedantic: `tr -d ' '`? :)

**Review by jean-edouard — COMMENTED** [review]
Looks good, just have a few questions in-line, mostly nits.

File: cluster-provision/k8s/1.19/node01.sh (hunk 6 above)
Comment (jordigilh) [reply cluster-provision/k8s/1.19/node01.sh:6]: Yep, would have done the trick. Unfortunately these 2 lines are no longer there... they were casualties to this PR :smile: (or should be if I missed them in any of the other scripts :smile:  )

File: cluster-provision/gocli/cmd/run.go (hunk 5 above)
Comment (jordigilh) [reply cluster-provision/gocli/cmd/run.go:795]: Sure, I'll put it back :smile:

File: cluster-provision/gocli/cmd/run.go (hunk 3 above)
Comment (jordigilh) [reply cluster-provision/gocli/cmd/run.go:523]: I wanted to leave the structure there in case it was required to support other sizes... (I initially planned to do so). I can definitely trim the code.

File: cluster-provision/k8s/1.19/node01.sh (hunk 6 above)
Comment (jean-edouard) [reply cluster-provision/k8s/1.19/node01.sh:6]: Facepalm, ignore me, time for the week-end! :)

File: cluster-provision/gocli/cmd/run.go (hunk 4 above)
Comment (jordigilh) [reply cluster-provision/gocli/cmd/run.go:741]: I don't think that the node will reboot during a run for realtime tests, but I can add the setting to `/etc/sysctl.d/`.

**Review by jean-edouard — APPROVED** [review]

File: cluster-provision/gocli/cmd/run.go (hunk 4 above)
Comment (jordigilh) [reply cluster-provision/gocli/cmd/run.go]: Ok done: The value is now stored in `/etc/sysctl.d/realtime.conf` and loaded with `sysctl --system`. This change only applies  to k8s-1.22 since the functional tests will only run with this version onwards.

**Comment by jordigilh** [comment]
@jean-edouard can I get another `lgtm`? Thanks! :smile:

File: cluster-provision/centos8/scripts/kernel.args (hunk 1 above)
Comment (rmohr) [reply cluster-provision/centos8/scripts/kernel.args]: I meant converging all `2m` settings into one single parameter and not adding a boolean. So that `--hugepages-2m=64` would add both kernel parameters. But happy if it works for you as is.

Comment (jordigilh) [reply cluster-provision/centos8/scripts/kernel.args]: Makes sense and it looks cleaner (1 flag vs 2 flags). I'll raise a new PR to adjust to this suggestion. Should be painless :smile:

Comment (vladikr) [reply cluster-provision/centos8/scripts/kernel.args]: We will still need another parameter if we will need to support 1G pages, won't we?

Comment (jordigilh) [reply cluster-provision/centos8/scripts/kernel.args]: That's true, but I wonder if that case is coming anytime soon (disclaimer: I don't know). If there is no plans in the horizon to support other than 2MB sizes for functional testing, it might be a case of over engineering.
Otherwise, I'm happy to leave the structure in the code to support other sizes :smile:
